File: backend/app/schemas/watcher.py
"""Pydantic schemas for Watcher"""
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime


class WatcherBase(BaseModel):
    """Base watcher schema"""
    name: str = Field(..., min_length=1, max_length=255)
    url: str = Field(..., min_length=1)
    method: str = Field(default="GET", pattern="^(GET|POST|PUT|PATCH|DELETE|HEAD|OPTIONS)$")
    headers: Optional[dict] = Field(default_factory=dict)
    body: Optional[str] = None
    
    # Content and execution settings
    content_type: str = Field(default="auto", pattern="^(auto|text|json|html|xml|image|pdf)$")
    execution_mode: str = Field(default="scheduled", pattern="^(scheduled|manual|both)$")
    watch_interval: Optional[int] = Field(default=None, gt=0)  # Required for scheduled mode
    is_active: bool = Field(default=True)
    
    # Cookie settings
    save_cookies: bool = Field(default=False)
    use_cookies: bool = Field(default=False)
    cookie_watcher_id: Optional[int] = None
    
    # Change detection
    comparison_mode: str = Field(default="hash", pattern="^(hash|content_aware|disabled)$")


class WatcherCreate(WatcherBase):
    """Schema for creating a watcher"""
    pass


class WatcherUpdate(BaseModel):
    """Schema for updating a watcher"""
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    url: Optional[str] = Field(None, min_length=1)
    method: Optional[str] = Field(None, pattern="^(GET|POST|PUT|PATCH|DELETE|HEAD|OPTIONS)$")
    headers: Optional[dict] = None
    body: Optional[str] = None
    
    # Content and execution settings
    content_type: Optional[str] = Field(None, pattern="^(auto|text|json|html|xml|image|pdf)$")
    execution_mode: Optional[str] = Field(None, pattern="^(scheduled|manual|both)$")
    watch_interval: Optional[int] = Field(None, gt=0)
    is_active: Optional[bool] = None
    
    # Cookie settings
    save_cookies: Optional[bool] = None
    use_cookies: Optional[bool] = None
    cookie_watcher_id: Optional[int] = None
    
    # Change detection
    comparison_mode: Optional[str] = Field(None, pattern="^(hash|content_aware|disabled)$")


class WatcherInDB(WatcherBase):
    """Schema for watcher in database"""
    id: int
    created_at: datetime
    updated_at: datetime
    last_checked_at: Optional[datetime] = None
    last_changed_at: Optional[datetime] = None
    status: str = "pending"
    error_message: Optional[str] = None
    check_count: int = 0
    change_count: int = 0

    class Config:
        from_attributes = True


class WatcherResponse(WatcherInDB):
    """Schema for watcher API response"""
    pass


class WatcherListResponse(BaseModel):
    """Simplified watcher schema for listing"""
    id: int
    name: str
    url: str
    execution_mode: str
    is_active: bool
    status: str
    check_count: int
    change_count: int
    last_checked_at: Optional[datetime] = None
    last_changed_at: Optional[datetime] = None
    created_at: datetime

    class Config:
        from_attributes = True


class WatcherWithStats(WatcherInDB):
    """Watcher with additional statistics"""
    recent_changes: int = 0
    success_rate: float = 0.0


class WatcherStatistics(BaseModel):
    """Statistics for watchers"""
    total_watchers: int
    active_watchers: int
    inactive_watchers: int
    total_checks: int
    total_changes: int
    by_execution_mode: dict
    by_status: dict
    by_content_type: dict

File: frontend/src/types/watcher.ts
export enum ExecutionMode {
  SCHEDULED = "scheduled",
  MANUAL = "manual",
  BOTH = "both"
}

export enum ContentType {
  AUTO = "auto",
  TEXT = "text",
  JSON = "json",
  HTML = "html",
  XML = "xml",
  IMAGE = "image",
  PDF = "pdf"
}

export interface Watcher {
  id: number;
  name: string;
  url: string;
  method: string;
  headers?: Record<string, string>;
  body?: string;
  content_type: ContentType;
  execution_mode: ExecutionMode;
  watch_interval?: number;
  is_active: boolean;
  save_cookies: boolean;
  use_cookies: boolean;
  cookie_watcher_id?: number;
  comparison_mode: string;
  status: string;
  error_message?: string;
  check_count: number;
  change_count: number;
  created_at: string;
  updated_at: string;
  last_checked_at?: string;
  last_changed_at?: string;
}

export interface WatcherCreate {
  name: string;
  url: string;
  method?: string;
  headers?: Record<string, string>;
  body?: string;
  content_type?: ContentType;
  execution_mode?: ExecutionMode;
  watch_interval?: number;
  is_active?: boolean;
  save_cookies?: boolean;
  use_cookies?: boolean;
  cookie_watcher_id?: number;
  comparison_mode?: string;
}

export interface WatcherUpdate {
  name?: string;
  url?: string;
  method?: string;
  headers?: Record<string, string>;
  body?: string;
  content_type?: ContentType;
  execution_mode?: ExecutionMode;
  watch_interval?: number;
  is_active?: boolean;
  save_cookies?: boolean;
  use_cookies?: boolean;
  cookie_watcher_id?: number;
  comparison_mode?: string;
}

export interface WatcherStatistics {
  total_watchers: number;
  active_watchers: number;
  inactive_watchers: number;
  total_checks: number;
  total_changes: number;
  by_execution_mode: Record<string, number>;
  by_status: Record<string, number>;
  by_content_type: Record<string, number>;
}

export interface WatcherExecutionResult {
  status: string;
  status_code?: number;
  response_body?: string;
  response_headers?: Record<string, string>;
  cookies_saved?: number;
  cookies_used?: number;
  error?: string;
}
